src/NATS.Client.Core/Commands/ProtocolWriter.cs

```diff
@@ -48,70 +48,131 @@ public void WritePong()
 
     // https://docs.nats.io/reference/reference-protocols/nats-protocol#pub
     // PUB <subject> [reply-to] <#bytes>\r\n[payload]\r\n
-    public void WritePublish(string subject, string? replyTo, NatsHeaders? headers, ReadOnlySequence<byte> payload)
+    // or
+    // https://docs.nats.io/reference/reference-protocols/nats-protocol#hpub
+    // HPUB <subject> [reply-to] <#header bytes> <#total bytes>\r\n[headers]\r\n\r\n[payload]\r\n
+    //
+    // returns the number of bytes that should be skipped when writing to the wire
+    public int WritePublish<T>(string subject, T? value, NatsHeaders? headers, string? replyTo, INatsSerialize<T> serializer)
```

Comment:
Have mixed feelings about this, would need to check what else has changed.

Main reason being, it looks like we've lowered a generic call further down (i.e. this is now `<T>` rather than `ReadOnlySequence<byte>`) and that may have implications around codegen for struct types.